cldfbench_doreco.py

```diff
@@ -15,21 +20,46 @@ def cldf_specs(self):  # A dataset must declare all CLDF sets it creates.
                 module='StructureDataset'
                 )
 
+    def cmd_download(self, args):
+        self.raw_dir.download(
+            'https://sharedocs.huma-num.fr/wl/?id=s947TcfRfDZR643QURdyncdku4EmeKyb&fmode=download',
+            'languages.csv')
+        self.raw_dir.download(
+            'https://sharedocs.huma-num.fr/wl/?id=xqkUR3WoFOKB8cRb0MdHYJxIDEXEXlVG&fmode=download',
+            'sources.bib')
+
+        for row in self.raw_dir.read_csv('languages.csv', dicts=True):
+            if 'ND' not in row['Annotation license']:
+                print(row['Glottocode'], row['DOI'])
+                cid = urllib.parse.quote(row['DOI'], safe='')
+                for f in requests.get("{}datas/{}?metadata-format=dc".format(NAKALA_API, cid)).json()['files']:
+                    for s in ['_wd.csv', '_ph.csv', '_metadata.csv', '_gloss-abbreviations.csv']:
+                        if f['name'].endswith(s):
+                            print(f['name'], f['sha1'])
+                            self.raw_dir.download('{}data/{}/{}'.format(NAKALA_API, cid, f['sha1']), '{}{}'.format(row['Glottocode'], s))
+                            break
+
+    def iter_rows(self, pattern):
+        for p in sorted(self.raw_dir.glob(pattern), key=lambda pp: p.name):
+            for row in self.raw_dir.read_csv(p.name, dicts=True):
+                row.setdefault('Glottocode', p.name.partition('_')[0])
+                if '_wd' in pattern or ('_ph' in pattern):
+                    # doreco-mb-algn col missing in some files of _ph
+                    row.setdefault('doreco-mb-algn', '')
+                yield row
+
     def cmd_makecldf(self, args):
         self.create_schema(args.writer.cldf)
 
-        sources = pybtex.database.parse_string(
-            'doreco_CITATIONS.bib', bib_format='bibtex'
-            )
-        args.writer.cldf.add_sources(sources)
+        args.writer.cldf.add_sources(pybtex.database.parse_string(
+            self.raw_dir.joinpath('sources.bib').read_text(encoding='utf8'),
+            bib_format='bibtex',
+        ))
         args.log.info("added sources")
 
-        for row in self.etc_dir.read_csv(
-            'doreco_languages_metadata.csv',
-            dicts=True,
-            ):
+        for row in self.raw_dir.read_csv('languages.csv', dicts=True):
             args.writer.objects["LanguageTable"].append({
-                "ID": row["Language"],
+                "ID": row["Glottocode"],
```

Comment:
Glottocodes make for "better" identifiers in the sense that they contain only tightly controlled characters, suitable in all sorts of contexts (HTL, URL parameters, etc.)

Reply:
There is also a row `mc-zero` that is missing to some datasets. Not quite sure what it refers to, but it should probably be added as well, right? Similar to `doreco-mb-algn`

Reply:
This refers to those lines: https://github.com/cldf-datasets/doreco/blob/robert/cldfbench_doreco.py#L46-L48